### PR TITLE
refactor(Core/ScalingStatValue): Allow use larger statistics on items

### PR DESCRIPTION
**Pull request by illunix**
I think players should be able to use items with larger statistics without core edits.

## HOW TO TEST THE CHANGES:

1. Recompile, clear client cache and log in game : Blizzlike items stats shouldn't had changed.
2. Recompile, clear client cache, create a custom item with stats > 32767 : new stats items shoudln't be limited to 32767.

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
@Shard-MW

File: data/sql/updates/db_world/2020_06_23_01.sql [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+FROM information_schema.COLUMNS
+WHERE TABLE_SCHEMA = DATABASE() AND TABLE_NAME = 'version_db_world' AND COLUMN_NAME = '2020_06_23_00';
+IF @COLEXISTS = 0 THEN LEAVE proc; END IF;
+START TRANSACTION;
+ALTER TABLE version_db_world CHANGE COLUMN 2020_06_23_00 2020_06_23_01 bit;
+SELECT sql_rev INTO OK FROM version_db_world WHERE sql_rev = '1588709577858892600'; IF OK <> 'FALSE' THEN LEAVE proc; END IF;
+--
+-- START UPDATING QUERIES
+--
+
```
Comment (Kitzunu) [inline data/sql/updates/db_world/2020_06_23_01.sql:16]: Please remove all this, it is automatically added by our bot when it is merged

File: data/sql/updates/db_world/2020_06_23_01.sql [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	CHANGE COLUMN `stat_value10` `stat_value10` INT NOT NULL DEFAULT 0 AFTER `stat_type10`;
+
+--
+-- END UPDATING QUERIES
+--
+COMMIT;
+END //
+DELIMITER ;
+CALL updateDb();
+DROP PROCEDURE IF EXISTS `updateDb`;
```
Comment (Kitzunu) [inline data/sql/updates/db_world/2020_06_23_01.sql:38]: This too

File: data/sql/updates/db_world/2020_06_24_01.sql [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+	CHANGE COLUMN `stat_value1` `stat_value1` INT NOT NULL DEFAULT 0 AFTER `stat_type1`,
+	CHANGE COLUMN `stat_value2` `stat_value2` INT NOT NULL DEFAULT 0 AFTER `stat_type2`,
+	CHANGE COLUMN `stat_value3` `stat_value3` INT NOT NULL DEFAULT 0 AFTER `stat_type3`,
+	CHANGE COLUMN `stat_value4` `stat_value4` INT NOT NULL DEFAULT 0 AFTER `stat_type4`,
+	CHANGE COLUMN `stat_value5` `stat_value5` INT NOT NULL DEFAULT 0 AFTER `stat_type5`,
+	CHANGE COLUMN `stat_value6` `stat_value6` INT NOT NULL DEFAULT 0 AFTER `stat_type6`,
+	CHANGE COLUMN `stat_value7` `stat_value7` INT NOT NULL DEFAULT 0 AFTER `stat_type7`,
+	CHANGE COLUMN `stat_value8` `stat_value8` INT NOT NULL DEFAULT 0 AFTER `stat_type8`,
+	CHANGE COLUMN `stat_value9` `stat_value9` INT NOT NULL DEFAULT 0 AFTER `stat_type9`,
+	CHANGE COLUMN `stat_value10` `stat_value10` INT NOT NULL DEFAULT 0 AFTER `stat_type10`;
```
Comment (FrancescoBorzi) [inline data/sql/updates/db_world/2020_06_24_01.sql:13]: this is wrong, SQL change should be submitted via pending SQL updates:

https://www.azerothcore.org/wiki/Dealing-with-SQL-files#how-to-create-an-sql-update-file

File: data/sql/base/db_world/item_template.sql [hunk 4]
```diff
@@ -38,25 +38,25 @@ CREATE TABLE `item_template`
...
+  `stat_value8` int(6) NOT NULL DEFAULT '0',
   `stat_type9` tinyint(3) unsigned NOT NULL DEFAULT '0',
-  `stat_value9` smallint(6) NOT NULL DEFAULT '0',
+  `stat_value9` int(6) NOT NULL DEFAULT '0',
   `stat_type10` tinyint(3) unsigned NOT NULL DEFAULT '0',
-  `stat_value10` smallint(6) NOT NULL DEFAULT '0',
+  `stat_value10` int(6) NOT NULL DEFAULT '0',
   `ScalingStatDistribution` smallint(6) NOT NULL DEFAULT '0',
   `ScalingStatValue` int(10) unsigned NOT NULL DEFAULT '0',
   `dmg_min1` float NOT NULL DEFAULT '0',
```
Comment (FrancescoBorzi) [inline data/sql/base/db_world/item_template.sql:62]: you should not edit the base files unless there is a specific need for that (e.g. the file can't be installed)

**Comment by Illunix** [comment]
Is that will be merged?

**Comment by Shard-MW** [comment]
Yes but we need you to resolve reviews first please

File: data/sql/updates/db_world/2020_06_24_01.sql (hunk 3 above)
Comment (Illunix) [reply data/sql/updates/db_world/2020_06_24_01.sql:13]: You can't change data type of column with update

**Comment by Kitzunu** [comment]
What is this?? https://github.com/azerothcore/azerothcore-wotlk/pull/3156/commits/71d7cecabca09e4af41ea13da4b7ce6e4148cf62

**Comment by Illunix** [comment]
> What is this?? [71d7cec](https://github.com/azerothcore/azerothcore-wotlk/commit/71d7cecabca09e4af41ea13da4b7ce6e4148cf62)

https://github.com/azerothcore/azerothcore-wotlk/pull/3156#discussion_r445335450 Changed data types to ``smallint``

**Comment by Kitzunu** [comment]
But you have 46,278 additions and 46,279 deletions

File: data/sql/updates/db_world/2020_06_24_01.sql (hunk 3 above)
Comment (Kitzunu) [reply data/sql/updates/db_world/2020_06_24_01.sql:13]: That is not what he meant.

You have created a file in the wrong directory. Please read the link.

**Comment by Kitzunu** [comment]
So I checked the changes, and you can resolve this two ways:
1. Revert ALL changes to item_template.sql (don't change, just revert)
2. Allow me to edit this PR and I can do it

**Comment by Illunix** [comment]
> So I checked the changes, and you can resolve this two ways:
> 
> 1. Revert ALL changes to item_template.sql (don't change, just revert)
> 2. Allow me to edit this PR and I can do it

I think you should be to edit this pr? @Kitzunu

**Comment by Hacaw** [comment]
I understood that the purpose of this issue is to raise the variable's data-type used for items to support bigger stats (seems like a requirement for a fun-server instant lvl 255 with cosmetic items), but smaller data-types might mean faster network transfers.

I think there are too many changes to be reviewed. For me this pull-request is not safe at all, even though I can easily test it, I would actually want to see what are the changes and to spend 1 week to look at those is a waste of time for everyone.
There are so many affected rows (just checked the [raw version](https://raw.githubusercontent.com/azerothcore/azerothcore-wotlk/02a77c916f88e372400cbb859783e1b5e3e2417e/data/sql/base/db_world/item_template.sql).

How can a manual process like a review should validate the recreation of the entire item_template table item by item?
How do I know items were forgot or messed up ?

I don't think this is the sql-way to do it, to manually add each item in the table. Wouldn't be better to use:

```
1. RENAME TABLE item_template to old_item_template;

2. CREATE TABLE new_item_template (with the structure you provided)

3. INSERT INTO new_item_template
SELECT * FROM old_item_template 

4. DROP TABLE old_item_template;

5. RENAME TABLE new_item_template to item_template;

```

@Helias @FrancescoBorzi @Yehonal what do you think about this PR ?

**Comment by Yehonal** [comment]
> think there are too many changes to be reviewed. For me

Actually, the network request doesn't change, I cannot check the size of the opcode chunk but if the opcode supports values higher than "smallint" I suppose the server is already sending an uint32 or something higher than smallint basically.

The cost over benefit here is that the database will relatively increase in size. It's not a problem actually, but the PR looks like to be still refactored, the author cannot edit the base and a proper migration (update) file must be created.

Moreover, we don't really need it but, if the PR doesn't change the official behavior and will be refactored properlt, it's good to have room for customization.

**Comment by Shard-MW** [comment]
> I understood that the purpose of this issue is to raise the variable's data-type used for items to support bigger stats (seems like a requirement for a fun-server instant lvl 255 with cosmetic items), but smaller data-types might mean faster network transfers.
> 
> I think there are too many changes to be reviewed. For me this pull-request is not safe at all, even though I can easily test it, I would actually want to see what are the changes and to spend 1 week to look at those is a waste of time for everyone.
> There are so many affected rows (just checked the [raw version](https://raw.githubusercontent.com/azerothcore/azerothcore-wotlk/02a77c916f88e372400cbb859783e1b5e3e2417e/data/sql/base/db_world/item_template.sql).
> 
> How can a manual process like a review should validate the recreation of the entire item_template table item by item?
> How do I know items were forgot or messed up ?
> 
> I don't think this is the sql-way to do it, to manually add each item in the table. Wouldn't be better to use:
> 
> ```
> 1. RENAME TABLE item_template to old_item_template;
> 
> 2. CREATE TABLE new_item_template (with the structure you provided)
> 
> 3. INSERT INTO new_item_template
> SELECT * FROM old_item_template 
> 
> 4. DROP TABLE old_item_template;
> 
> 5. RENAME TABLE new_item_template to item_template;
> ```
> 
> @Helias @FrancescoBorzi @Yehonal what do you think about this PR ?

I'm **sorry** but what's this ? It's like "How do we prove that 2 + 2 = 4 ?"
I don't understand why there can be "too many changes to be reviewed" since this PR should only change the datatype and the way the core read it. This patch exists since years now, and never caused any problem to a server...
Data inside `item_template` isn't changed at all, structure is. So we don't need a table copy + insert. We just need an `ALTER TABLE` + core fix (which is included in this WIP PR). 

> but smaller data-types might mean faster network transfers
Also this is true, but it does (only?) impact the server boot time and even in this case, it's a minor improvement... (+ a lot of running servers are using a local DB)

**Comment by Yehonal** [comment]
> > I understood that the purpose of this issue is to raise the variable's data-type used for items to support bigger stats (seems like a requirement for a fun-server instant lvl 255 with cosmetic items), but smaller data-types might mean faster network transfers.
> > I think there are too many changes to be reviewed. For me this pull-request is not safe at all, even though I can easily test it, I would actually want to see what are the changes and to spend 1 week to look at those is a waste of time for everyone.
> > There are so many affected rows (just checked the [raw version](https://raw.githubusercontent.com/azerothcore/azerothcore-wotlk/02a77c916f88e372400cbb859783e1b5e3e2417e/data/sql/base/db_world/item_template.sql).
> > How can a manual process like a review should validate the recreation of the entire item_template table item by item?
> > How do I know items were forgot or messed up ?
> > I don't think this is the sql-way to do it, to manually add each item in the table. Wouldn't be better to use:
> > ```
> > 1. RENAME TABLE item_template to old_item_template;
> > 
> > 2. CREATE TABLE new_item_template (with the structure you provided)
> > 
> > 3. INSERT INTO new_item_template
> > SELECT * FROM old_item_template 
> > 
> > 4. DROP TABLE old_item_template;
> > 
> > 5. RENAME TABLE new_item_template to item_template;
> > ```
> > 
> > 
> > @Helias @FrancescoBorzi @Yehonal what do you think about this PR ?
> 
> I'm **sorry**, I don't wanna sound mean but what's this ? It's like "How do we prove that 2 + 2 = 4 ?"
> I don't understand why there can be "too many changes to be reviewed" since this PR should only change the datatype and the way the core read it. This patch exists since years now, and never caused any problem to any server...
> Data inside `item_template` isn't changed at all, structure is. So we don't need a table copy + insert. We just need an `ALTER TABLE` + core fix (which is included in this WIP PR).
> 
> > but smaller data-types might mean faster network transfers
> 
> Also this is true, but it does (only?) impact the server boot time and even in this case, it's a minor improvement... (+ a lot of running servers are using a local DB)

I think he meant the changes on the base table, those must be reverted and use a proper update process. But the author didn't update the PR and I think it's useless to review/test for now since it's stale

**Review by Yehonal — CHANGES_REQUESTED** [review]
This PR cannot modify the base SQL and the update file must follow our procedure for the updates. Looks like it's stale since June, we need to wait for the author anyway.

**Pull request closed without merging**